lib/core/ingestion/es_sink.rb

```diff
@@ -11,49 +11,108 @@
   module Ingestion
     class EsSink
-      def initialize(index_name, request_pipeline, bulk_queue = Utility::BulkQueue.new)
+      def initialize(index_name, request_pipeline, bulk_queue = Utility::BulkQueue.new, max_allowed_document_size = 5 * 1024 * 1024)
         @client = Utility::EsClient.new(App::Config[:elasticsearch])
         @index_name = index_name
         @request_pipeline = request_pipeline
         @operation_queue = bulk_queue
+
+        @max_allowed_document_size = max_allowed_document_size
+
+        @queued = {
+          :indexed_document_count => 0,
+          :deleted_document_count => 0,
+          :indexed_document_volume => 0
+        }
+
+        @completed = {
+          :indexed_document_count => 0,
+          :deleted_document_count => 0,
+          :indexed_document_volume => 0
+        }
       end
 
-      def ingest(id, serialized_document)
-        index_op = serialize({ 'index' => { '_index' => index_name, '_id' => id } })
+      def ingest(document)
+        unless document&.any?
+          Utility::Logger.warn('Connector attempted to ingest an empty document, skipping')
+          return
+        end
+
+        id = document['id']
+        serialized_document = serialize(document)
+
+        document_size = serialized_document.bytesize
+
+        if @max_allowed_document_size > 0 && document_size > @max_allowed_document_size
+          Utility::Logger.warn("Connector attempted to ingest too large document with id=#{document['id']} [#{document_size}/#{@max_allowed_document_size}], skipping the document.")
+          return
+        end
+
+        index_op = serialize({ 'index' => { '_index' => @index_name, '_id' => id } })
 
         flush unless @operation_queue.will_fit?(index_op, serialized_document)
 
         @operation_queue.add(
           index_op,
           serialized_document
         )
+
+        @queued[:indexed_document_count] += 1
+        @queued[:indexed_document_volume] += document_size
+      end
+
+      def ingest_multiple(documents)
+        documents.each { |doc| ingest(doc) }
       end
 
-      def delete(doc_id)
-        delete_op = serialize({ 'delete' => { '_index' => index_name, '_id' => doc_id } })
+      def delete(id)
+        return if id.nil?
+
+        delete_op = serialize({ 'delete' => { '_index' => @index_name, '_id' => id } })
         flush unless @operation_queue.will_fit?(delete_op)
 
         @operation_queue.add(delete_op)
+        @queued[:deleted_document_count] += 1
+      end
+
+      def delete_multiple(ids)
+        ids.each { |id| delete(id) }
       end
 
       def flush
         data = @operation_queue.pop_all
         return if data.empty?
 
         @client.bulk(:body => data, :pipeline => @request_pipeline)
+
+        @completed[:indexed_document_count] += @queued[:indexed_document_count]
+        @completed[:deleted_document_count] += @queued[:deleted_document_count]
+        @completed[:indexed_document_volume] += @queued[:indexed_document_volume]
```

Comment:
Should we actually grab these values from the response to the `bulk` request? Hypothetically, some of these operations might fail, in which case we probably shouldn't count them as successes.

Reply:
That is possible, though right now the logic is such that if `bulk` fails, then we fail the sync fully.

Once we add ingestion resilience, this will have to change to the logic you described. We can do it now too, but I don't feel I have capacity for it currently

Reply:
It's definitely more accurate, but the [response of the Bulk API ](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-api-response-body) doesn't return the volume, we certainly can have an id-to-volume map, and calculate the the total volume based on the response, which I think it's a bit overkill.